### PR TITLE
RFC: Icons partials

### PR DESCRIPTION
**Pull request by asudoh**
Refs IBM/carbon-elements#10.

With this change, experimental mode starts using experimental icons (e.g. the experimental code snippet shows a different copy icon from the copy icon in dev env):

image

Staging link: https://carbon-dev-environment-hippest-flaccidness.mybluemix.net

#### Changelog

**New**

- A mechanism to convert icon data from `carbon-elements` to Handlebars partials
- A mechanism to register such Handlebars partials

**Changed**

- Replaced hard-coded `<svg>` with the Handlebars partials above, wherever matching one in our icons library is found

#### Testing / Reviewing

Testing should make sure no component is broken, especially the icons in them.

### CODE REVIEW

File: src/components/accordion/accordion.hbs [hunk 1]
```diff
@@ -2,9 +2,7 @@
   {{#each sections}}
     <li data-accordion-item class="{{@root.prefix}}--accordion__item">
       <button class="{{@root.prefix}}--accordion__heading" aria-expanded="false" aria-controls="{{paneId}}">
-        <svg class="{{@root.prefix}}--accordion__arrow" width="7" height="12" viewBox="0 0 7 12">
-          <path fill-rule="nonzero" d="M5.569 5.994L0 .726.687 0l6.336 5.994-6.335 6.002L0 11.27z" />
-        </svg>
+        {{> carbon-icon-chevron-right newContext class=(add @root.prefix "--accordion__arrow") }}
```
Comment (joshblack) [inline src/components/accordion/accordion.hbs]: Is there any way to simplify the way we can add classes for partials?

Also, is `newContext` required? If so, why?

File: tools/templates.js [hunk 2]
```diff
@@ -42,16 +44,112 @@ try {
   console.error('Failed to hook Fractal logger', err.stack); // eslint-disable-line no-console
 }
 
+/**
```
Comment (joshblack) [inline tools/templates.js:47]: These methods feel like custom work without tests that could break when updates are received. How could we guard against them breaking? Could we use some of the helpers from `@carbon/icon-helpers` for things like attributes or its `toString` method?

File: tools/templates.js [hunk 3]
```diff
@@ -42,16 +44,112 @@ try {
...
+        []
+      ),
+  };
+};
+
+/**
+ * @param {boolean} useElements `true` to use one from `carbon-elements`.
+ * @returns {Map<string, string>} A set of icons contents, keyed by hyphnated icon name.
+ */
+const getIconsPartials = useElements => {
```
Comment (joshblack) [inline tools/templates.js]: How do we know what the list of available partials are? In other words, how do we know the name of the icon that we want to use.

File: tools/templates.js [hunk 4]
```diff
@@ -103,6 +201,7 @@ const cache = {
    * Clears the content cache.
    */
   clear() {
+    delete require.cache[require.resolve('../demo/feature-flags.js')];
```
Comment (joshblack) [inline tools/templates.js:115]: Could we leave a comment explaining why this is necessary?

**Review by joshblack — COMMENTED** [review]
I'm concerned about tying more of this implementation to the environment, especially when it's an area of the codebase that there isn't a lot of information about and there aren't necessarily tests to help guard us against regressions or things that may happen unexpectedly.

In addition, using these partials feels a bit cumbersome, especially for class names. Is there an easy strategy for passing in attributes?

File: tools/templates.js (hunk 2 above)
Comment (asudoh) [reply tools/templates.js:47]: Thank you for bringing up `@carbon/icon-helpers` library. [`formatAttributes()`](https://github.com/IBM/carbon-elements/blob/v0.0.1-alpha.12/packages/icon-helpers/src/toString.js#L17-L25) is definitely something that can be used. Wrt [`toString()`](https://github.com/IBM/carbon-elements/blob/v0.0.1-alpha.12/packages/icon-helpers/src/toString.js#L6-L15), we need a way to inject `{{#if description}} aria-label="{{description}}"{{/if}}`/`{{#if description}}<title>{{description}}</title>{{/if}}`. Also test sounds a valid point, which makes me wonder if you have a plan to add tests in `@carbon/icon-helpers`, or are you planning a manual way to get `@carbon/icon` and `@carbon/icon-helpers` in sync?

File: tools/templates.js (hunk 3 above)
Comment (asudoh) [reply tools/templates.js]: The icon names are from `@carbon/icons`/`carbon-icons` named exports without size/`Glyph` names, in a hyphenated format, e.g. `carbon-icon-caret-right` corresponds to `CaretRightGlyph`. But I now see that `@carbon/icons` has a `name` property in the data, which was exactly what I wanted - Will make a change to use it.

File: tools/templates.js (hunk 4 above)
Comment (asudoh) [reply tools/templates.js:115]: Definitely - It's for avoiding stale import cache of feature flags after the theme switcher in dev env UI is clicked on.

File: src/components/accordion/accordion.hbs (hunk 1 above)
Comment (asudoh) [reply src/components/accordion/accordion.hbs]: Good point, an alternate approach is doing string interpolation in the partials (by e.g. [`{{replace class "{{@root.prefix}}" "bx"}}`](https://github.com/helpers/handlebars-helpers#replace)). It allows us to do `class="{{@root.prefix}}--accordion__arrow"`. That said, Handlebars does not seem to do standard string interpolation for partial args.

`newContext` is for preventing parent context passed along to the partial. Without `newContext`, I didn't see a way to cleanly distinguish partial args for additional attributes vs. parent context. Probably prefixing partial args is a good way for that. Another way is allowing limited set of attributes.

File: tools/templates.js (hunk 2 above)
Comment (joshblack) [reply tools/templates.js:47]: We have tests for `@carbon/icon-helpers` here: https://github.com/IBM/carbon-elements/tree/master/packages/icon-helpers/src/__tests__ feel free to suggest updates, changes, etc. 👍

**Comment by asudoh** [comment]
Made the following changes:

* String interpolation of CSS class prefix inside partials
* Refer to name property in icon data to make name determination logic more robust
* Avoid usage of `useContext` by using `attr-` prefix for attributes
* Utilize `@carbon/icon-helpers`
* Comment for module cache cleanup

Thanks @joshblack for all your comments!

File: tools/templates.js (hunk 2 above)
Comment (asudoh) [reply tools/templates.js:47]: Thanks @joshblack to point to where the test is - Looks good and too bad that I couldn't find it earlier.

**Comment by shinytoyrobots** [comment]
I believe that as we moved forward with #1440 that this is currently unnecessary, so am closing. Feel free to reopen with comment if that's incorrect.

**Comment by asudoh** [comment]
@shinytoyrobots This PR is for improving #1440, from the following perspectives:

* Avoid additional Handlebars helpers dependency (This library tries to limit Handlebars dependencies to well-known ones so future effort of shipping Handlebars templates won't end up with "missing helpers" error)
* Foster a mechanism in dev env in this repo to easily swap stable/experimental icons (So single `{{> carbon-icon-icon-name }}` call work both for stable/experimental icons)

Does this make sense to you...? Please don't hesitate to ping me if you have any questions. Thanks!

**Comment by joshblack** [comment]
For things like shipping handlebar templates, I believe we would have to spin up a standalone effort versus using what exists already in this environment as the work in `carbon-components` would be considered a dev-only environment, correct?

For things like easily swapping, the goal would be to switch wholesale to one approach. This period in-between should be ephemeral while we are working alongside design to figure out what's the correct behavior here.

**Comment by asudoh** [comment]
Good questions/thoughts @joshblack!

> For things like shipping handlebar templates, I believe we would have to spin up a standalone effort versus using what exists already in this environment as the work in carbon-components would be considered a dev-only environment, correct?

I imagined the Handlebars template we have now would evolved to shippable ones. Given you seem to have imagined a whole new set of shippable templates, good to see why you thought so.

> For things like easily swapping, the goal would be to switch wholesale to one approach. This period in-between should be ephemeral while we are working alongside design to figure out what's the correct behavior here.

You are right that `carbon-icons` and `@carbon/icons` co-existence is temporary, imagining a half year. I don't want to strike out requirements of lower-level library just with the state of what we have right now in our plate. The requirements I brought up here is more from how to make the lower-level library and the system more flexible, which is essential for better adoption/consumption.

**Comment by asudoh** [comment]
Made updates to match @joshblack's latest code upon designers' icons spec (thanks @joshblack for the work!). Also added some test against the code generating/registering Handlebars partials: https://github.com/IBM/carbon-components/blob/5004249a19b1f65540d7ac4c58f1706972a7e3c5/tests/partials-test.js

**Comment by joshblack** [comment]
_PR Reviews January 11th_

We took a look at this as part of our PR Reviews meeting on January 11th

Notes:

- Implementation could help out with conditional logic
- Some of the templates became harder to read due to obfuscation of what is going on
- Value decreases over time, might be worth looking at if we do this type of thing again

Summary: code looks fine but it's not a priority. Closing this for now, would be great as a reference in the future if we do this type of thing again 👍

**Pull request closed without merging**